### ISSUE TITLE
Does func support TLS-enabled docker hosts?

### ISSUE THREAD

**Issue by riverar** [knative/func#1627]
When attempting to deploy a function to a remote Docker daemon, func attempts to send an HTTP request to an HTTPS-only endpoint. Specifically, the following error is output:

```
> func version
v1.9.1

> func deploy -v --registry xxx
Building function image
Pulling image gcr.io/paketo-buildpacks/builder:base
Error: failed to fetch builder image 'gcr.io/paketo-buildpacks/builder:base': Error response from daemon: Client sent an HTTP request to an HTTPS server.
```

My `DOCKER_HOST` is set to `tcp://redacted.host:2376` and `DOCKER_TLS_VERIFY` is set. I've verified normal docker commands (e.g. `docker ls` work correctly.)

Windows 10.0.25314.1010

**Comment by lance** [comment]
@riverar what happens when you `docker pull gcr.io/paketo-buildpacks/builder:base`?

**Comment by riverar** [comment]
It completes without error.

**Comment by riverar** [comment]
Hm. I replied too quickly. It previously ran but that's erroring out too.

```
> docker pull gcr.io/paketo-buildpacks/builder:base
Error response from daemon: Client sent an HTTP request to an HTTPS server
```

Let me go check on the daemon. This is looking like operator error, apologies.

**Comment by matejvasek** [comment]
I'll look at this.

**Comment by riverar** [comment]
OK, I have gotten my system back into the originally reported state.

> docker pull gcr.io/paketo-buildpacks/builder:base
...
f6dca84f435e: Pull complete
72ad9888618d: Pull complete
4f4fb700ef54: Pull complete
Digest: sha256:d7870b73e331d2ba58de0d642a40777284def3c75bd4d90268202b2538028755
Status: Downloaded newer image for gcr.io/paketo-buildpacks/builder:base
gcr.io/paketo-buildpacks/builder:base

> kn-func deploy -v --registry xxx
Building function image
Pulling image gcr.io/paketo-buildpacks/builder:base
Error: failed to fetch builder image 'gcr.io/paketo-buildpacks/builder:base': Error response from daemon: Client sent an HTTP request to an HTTPS server.
```

Happy to spin up a machine or help with this live/over a call if you don't have Windows resources readily available. Thanks!

**Comment by matejvasek** [comment]
@riverar does `func build` work?

**Comment by matejvasek** [comment]
This is bug in `func` I think.
RE: `DOCKER_TLS_VERIFY` -- this only tell to verify or not verify server certificate.

**Comment by matejvasek** [comment]
@riverar if you are on Linux you can workaround this by using `socat`.

**Comment by matejvasek** [comment]
`func` also support `ssh` connections

**Comment by riverar** [comment]
> @riverar does `func build` work?

Sadly no.

```
kn-func build

   Building function image

2023/03/15 11:06:58.646235 DEBUG:  Pulling image gcr.io/paketo-buildpacks/builder:base

Error: failed to build the function: failed to fetch builder image 'gcr.io/paketo-buildpacks/builder:base': Error response from daemon: Client sent an HTTP request to an HTTPS server.    
```

Thanks for confirming. `socat tcp-listen:2372,fork,reuseaddr openssl:0.0.0.0:2376` is an easy workaround for private use.

**Comment by matejvasek** [comment]
It seems that the `pack` CLI also suffers from this too.
Docker client created by `client.NewClientWithOpts(client.FromEnv)` doesn't seems to work properly with tls.
The Docker CLI must do some more sophisticated initialization.

**Comment by matejvasek** [comment]
TBH I would expect `client.NewClientWithOpts(client.FromEnv)` to initialise everything all right.

**Comment by riverar** [comment]
Agreed. Am I reading this correctly? They only attempt to configure TLS if my `DOCKER_HOST` is configured without a scheme? https://github.com/docker/docker-ce/blob/master/components/engine/client/client.go#L141-L153

**Comment by matejvasek** [comment]
From my testing it seems that you need to set `DOCKER_TLS_VERIFY`. It may be true or false, but it must be set. Then TLS is used with TCP.

**Comment by matejvasek** [comment]
Also existence of `~/.docker/ca.pem` may play a part.

**Comment by matejvasek** [comment]
https://github.com/docker/cli/blob/4595ce588cd19879a33f32150c33a6bc3571b0a8/cli/flags/options.go#L94-L100

**Comment by riverar** [comment]
Yeah I have `DOCKER_TLS_VERIFY=1` already and a set of CA certificates on disk (docker CLI is working).

**Comment by matejvasek** [comment]
```go
t := &http.Transport{}
t.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
dockerClient, err = client.NewClientWithOpts(client.FromEnv, client.WithHTTPClient(&http.Client{Transport: t}))
```
this for instance makes tls work

**Comment by matejvasek** [comment]
Basically if TLS is explicitly set (envvar or CLI flag), the docker CLI calls `client.NewClientWithOpts()` with `client.WithHTTPClient()` with http clien such that its TLS config is not `nil`.

**Comment by riverar** [comment]
Probably obvious but if you're working on a PR, I kindly suggest dropping `InsecureSkipVerify: true` 🙃

**Comment by matejvasek** [comment]
Hmm please try setting `DOCKER_CERT_PATH`  to `$HOME/.docker/`.

**Comment by matejvasek** [comment]
https://github.com/knative/func/blob/94e49eb3cdb3a20a08574b9757be86fdfd16abdb/vendor/github.com/docker/docker/client/options.go#L27-L43

**Comment by matejvasek** [comment]
Looks `client.FromEnv` can do the job, but you also need to set `DOCKER_CERT_PATH`.

**Comment by matejvasek** [comment]
So works as expected :rofl:  ?

**Comment by matejvasek** [comment]
hmm its not working as I thought it would.

**Comment by matejvasek** [comment]
It is supposed to work when both client&server are verified. So `client.FromEnv` is not enough :cry: .

**Comment by matejvasek** [comment]
@riverar out of curiosity what is your set up? Why are you using tcp+tls?

**Comment by riverar** [comment]
@matejvasek I'm using a Windows client with a remote Linux-based container host. I'm configured for TLS because my domain name is [HSTS preloaded](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security) which generally means HTTP is unavailable.

**Comment by matejvasek** [comment]
I see. For such a set up `ssh` might be a better fit since it also offers authentication.
It is also possible authenticate with tcp+tls if you set up client certificates.

**Comment by matejvasek** [comment]
Fixed by #1629
